File: MoonReader/Core/Parsers/MDParser.swift
//
//  MDParser.swift
//  MoonReader
//
//  Markdown Parser - tương đương Md.java trong Android
//

import Foundation

class MDParser: BaseBookParser {
    override func extractMetadata(fileURL: URL) throws -> BookMetadata {
        let filename = fileURL.lastPathComponent
        let title = (filename as NSString).deletingPathExtension
        
        return BookMetadata(
            title: title,
            author: "",
            description: nil,
            publisher: nil,
            isbn: nil,
            publishDate: nil,
            language: nil,
            coverImage: nil
        )
    }
    
    override func extractChapters(fileURL: URL) throws -> [Chapter] {
        guard let content = try? String(contentsOf: fileURL, encoding: .utf8) else {
            throw BookParserError.corruptedFile
        }
        
        // Split by headers (# ## ###)
        let lines = content.components(separatedBy: .newlines)
        var chapters: [Chapter] = []
        var currentChapter: (title: String, content: [String])?
        
        for line in lines {
            if line.hasPrefix("# ") {
                // Save previous chapter
                if let chapter = currentChapter {
                    let chapterContent = chapter.content.joined(separator: "\n")
                    chapters.append(Chapter(
                        id: "chapter_\(chapters.count)",
                        title: chapter.title,
                        content: chapterContent,
                        order: chapters.count,
                        filePath: nil
                    ))
                }
                // Start new chapter
                let title = String(line.dropFirst(2)).trimmingCharacters(in: .whitespaces)
                currentChapter = (title: title, content: [])
            } else if let chapter = currentChapter {
                currentChapter?.content.append(line)
            }
        }
        
        // Add last chapter
        if let chapter = currentChapter {
            let chapterContent = chapter.content.joined(separator: "\n")
            chapters.append(Chapter(
                id: "chapter_\(chapters.count)",
                title: chapter.title,
                content: chapterContent,
                order: chapters.count,
                filePath: nil
            ))
        }
        
        // If no chapters found, create one with all content
        if chapters.isEmpty {
            chapters.append(Chapter(
                id: "chapter_0",
                title: "Nội dung",
                content: content,
                order: 0,
                filePath: nil
            ))
        }
        
        return chapters
    }
}

File: MoonReaderRN/src/core/parsers/MDParser.ts
/**
 * Markdown Parser
 * Parse Markdown files and split by headers
 */

import { BaseBookParser } from './BaseBookParser';
import { BookMetadata, Chapter } from './types';

export class MDParser extends BaseBookParser {
  /**
   * Extract metadata from MD file
   */
  async extractMetadata(fileUri: string): Promise<BookMetadata> {
    const title = this.getFilenameWithoutExtension(fileUri);

    return {
      title,
      author: '',
      description: undefined,
      publisher: undefined,
      isbn: undefined,
      publishDate: undefined,
      language: undefined,
      coverImage: undefined,
    };
  }

  /**
   * Extract chapters from MD file
   * Splits by headers (# ## ###)
   */
  async extractChapters(fileUri: string): Promise<Chapter[]> {
    const content = await this.readFileAsString(fileUri);
    const lines = content.split('\n');

    const chapters: Chapter[] = [];
    let currentChapter: { title: string; content: string[] } | null = null;

    for (const line of lines) {
      // Check for header (starts with #)
      if (line.trim().match(/^#{1,3}\s+/)) {
        // Save previous chapter
        if (currentChapter) {
          const chapterContent = currentChapter.content.join('\n').trim();
          if (chapterContent.length > 0) {
            chapters.push({
              id: `chapter_${chapters.length}`,
              title: currentChapter.title,
              content: chapterContent,
              order: chapters.length,
              filePath: undefined,
            });
          }
        }

        // Start new chapter
        const title = line.replace(/^#{1,3}\s+/, '').trim();
        currentChapter = { title, content: [] };
      } else if (currentChapter) {
        // Add line to current chapter
        currentChapter.content.push(line);
      }
    }

    // Add last chapter
    if (currentChapter) {
      const chapterContent = currentChapter.content.join('\n').trim();
      if (chapterContent.length > 0) {
        chapters.push({
          id: `chapter_${chapters.length}`,
          title: currentChapter.title,
          content: chapterContent,
          order: chapters.length,
          filePath: undefined,
        });
      }
    }

    // If no chapters found, create one with all content
    if (chapters.length === 0) {
      chapters.push({
        id: 'chapter_0',
        title: 'Nội dung',
        content: content,
        order: 0,
        filePath: undefined,
      });
    }

    return chapters;
  }
}

